lib/consensus.js

```diff
@@ -1,11 +1,105 @@
 const { isValidTarget } = require('@dashevo/dark-gravity-wave');
+
 const utils = require('./utils');
 
-module.exports = {
+const MIN_DGW_HEADERS = 24;
+const MIN_TIMESTAMP_HEADERS = 11;
+
+function getDgwHeaders(headers, index) {
+  if (index >= MIN_DGW_HEADERS && headers.length >= MIN_DGW_HEADERS) {
+    return headers.slice(index - MIN_DGW_HEADERS, index);
+  }
+  return headers;
+}
+
+function getTimestampHeaders(headers, index) {
+  if (index > MIN_TIMESTAMP_HEADERS && headers.length > MIN_TIMESTAMP_HEADERS) {
+    return headers.slice(index - MIN_TIMESTAMP_HEADERS, index);
+  }
+  return [];
+}
+
+function isParentChild(parentHeader, childHeader) {
+  return utils.normalizeHeader(parentHeader).hash === utils.normalizeHeader(childHeader).prevHash;
+}
+
+function getTimestampAverage(headers) {
+  if (headers.length <= MIN_TIMESTAMP_HEADERS) return 0;
+  const timestamps = headers.slice(
+    headers.length - MIN_TIMESTAMP_HEADERS,
+  ).map(h => h.time);
+  return timestamps.reduce((total, amount, index, array) => {
+    total += amount; // eslint-disable-line no-param-reassign
+    if (index === array.length - 1) {
+      return total / array.length;
+    }
+    return total;
+  });
+}
+
+function isValidTimestamp(header, previousHeaders) {
+  if (previousHeaders.length <= MIN_TIMESTAMP_HEADERS) return true;
+  const avg = getTimestampAverage(previousHeaders);
+  return avg < header.time;
+}
+
+function validateHeaderChunk(chunk) {
+  const normalizedHeaders = chunk.map(header => utils.normalizeHeader(header));
+
+  // verifying timestamps if possible (ts > avg of previous 11 timestamps)
+  const invalidTimestamp = normalizedHeaders.filter(
+    (header, index, headers) => !isValidTimestamp(
+      header, getTimestampHeaders(headers, index),
+    ),
+  );
 
-  isValidBlockHeader(dgwHeaders, newHeader) {
-    return newHeader.validProofOfWork()
-      && newHeader.validTimestamp()
-      && isValidTarget(newHeader.bits, dgwHeaders.map(h => utils.getDgwBlock(h)));
-  },
+  // verifying POW
+  const invalidPOW = normalizedHeaders.filter(header => !header.validProofOfWork());
+
+  // verifying difficulty targets
+  const invalidTarget = normalizedHeaders.filter(
+    (header, index, headers) => !isValidTarget(
+      header.bits, getDgwHeaders(headers, index).map(h => utils.getDgwBlock(h)),
+    ),
+  );
+
+  // verifying parent child connections (chain consistency)
+  const badConnections = normalizedHeaders.filter((header, index, headers) => {
+    if (index > 0 || index < headers.length) {
+      return !isParentChild(header, headers[index + 1]);
+    }
+    return false;
+  });
+
+  const badHeaders = invalidTimestamp.concat(invalidPOW, invalidTarget, badConnections);
+
+  return { valid: badHeaders.length <= 0, badHeaders };
+}
+
+function validateHeaderChunks(chunks) {
```

Comment:
I was thinking, maybe it is better to not have `chunks`? My point is, SPV chain itself is in-memory storage and removes duplicates anyway. And it make sense to build retry logic around `addHeaders`. Please, let me know what you think.

Reply:
Problem is that `addHeaders` which uses `addHeader` to add a header to the branches makes use of `isValidBlockHeader` which can only check the validity of a freshly mined header because of the timestamp checking. That timestamp check could be removed or adapted of course. I went for this chunk based validation, because I didn't immediately see an easy way to know exactly which chunk was bad when waiting for the validation to happen in `addHeader`. That's why I would like to keep the chunk validation separated and re-request those which are bad with the exact same parameters that were used in the first call.

Reply:
You are completely right @jawid-h. We can add the missing validatiion easily in `addHeaders` which keeps everything much more clean. Closing this in favour of
https://github.com/dashevo/dash-spv/pull/18